airflow/providers/databricks/plugins/databricks_workflow.py

```diff
@@ -0,0 +1,440 @@
+    :return: The DagRun object associated with the specified DAG and run_id.
+    """
+    return session.query(DagRun).filter(DagRun.dag_id == dag.dag_id, DagRun.run_id == run_id).first()
+
+
+@provide_session
+def _clear_task_instances(dag_id: str, run_id: str, task_ids: list[str], log: logging.Logger, session=None):
+    dag = airflow_app.dag_bag.get_dag(dag_id)
+    log.debug("task_ids to clear", str(task_ids))
+    dr: DagRun = _get_dagrun(dag, run_id, session=session)
+    tis_to_clear = [ti for ti in dr.get_task_instances() if _get_databricks_task_id(ti) in task_ids]
+    clear_task_instances(tis_to_clear, session)
+
+
+def _repair_task(
+    databricks_conn_id: str,
+    databricks_run_id: int,
+    tasks_to_repair: list[str],
+    log: logging.Logger,
+) -> int:
+    """
+    Repair a Databricks task using the Databricks API.
+
+    This function allows the Airflow retry function to create a repair job for Databricks.
+    It uses the Databricks API to get the latest repair ID before sending the repair query.
+
+    :param databricks_conn_id: The Databricks connection ID.
+    :param databricks_run_id: The Databricks run ID.
+    :param tasks_to_repair: A list of Databricks task IDs to repair.
+    :return: None
+    """
+    hook = DatabricksHook(databricks_conn_id=databricks_conn_id)
+
+    repair_history_id = hook.get_latest_repair_id(databricks_run_id)
+    log.debug("Latest repair ID is %s", repair_history_id)
+    log.debug(
+        "Sending repair query for tasks %s on run %s",
+        tasks_to_repair,
+        databricks_run_id,
+    )
+
+    repair_json = {
+        "run_id": databricks_run_id,
+        "latest_repair_id": repair_history_id,
+        "rerun_tasks": tasks_to_repair,
+    }
+
+    return hook.repair_run(repair_json)
+
+
+def get_launch_task_id(task_group: TaskGroup) -> str:
+    """
+    Retrieve the launch task ID from the current task group or a parent task group, recursively.
+
+    :param task_group: Task Group to be inspected
+    :return: launch Task ID
+    """
+    try:
+        launch_task_id = task_group.get_child_by_label("launch").task_id  # type: ignore[attr-defined]
+    except KeyError as e:
+        if not task_group.parent_group:
+            raise AirflowException("No launch task can be found in the task group.") from e
+        launch_task_id = get_launch_task_id(task_group.parent_group)
+
+    return launch_task_id
+
+
+def _get_launch_task_key(current_task_key: TaskInstanceKey, task_id: str) -> TaskInstanceKey:
+    """
+    Return the task key for the launch task.
+
+    This allows us to gather databricks Metadata even if the current task has failed (since tasks only
+    create xcom values if they succeed).
+
+    :param current_task_key: The task key for the current task.
+    :param task_id: The task ID for the current task.
+    :return: The task key for the launch task.
+    """
+    if task_id:
```

Comment:
I just want to confirm what we are actually checking here. Whether the task_id is `""` or not?

Reply:
This one tries to add the task_id to the provided task instance key(current_task_key) if task id is given. We need to revisit/refactor this for clarity, but for now to keep the interface consistent between the migrating provider, I am keeping it as is.